panel/custom.py

```diff
@@ -0,0 +1,387 @@
+          Name of the DOM event to add an event listener to.
+        callback: callable
+          A callable which will be given the DOMEvent object.
+        """
+        self._event_callbacks[event].append(callback)
+
+
+class JSComponent(ReactiveESM):
+    '''
+    The `JSComponent` allows you to create custom Panel components
+    using Javascript and CSS without the complexities of
+    Javascript build tools.
+
+    A `JSComponent` subclass provides bi-directional syncing of its
+    parameters with arbitrary HTML elements, attributes and
+    properties. The key part of the subclass is the `_esm`
+    variable. Use this to define a `render` function as shown in the
+    example below.
+
+    import panel as pn
+    import param
+
+    pn.extension()
+
+    class CounterButton(JSComponent):
+
+        value = param.Integer()
+
+        _esm = """
+        export function render({ data }) {
+            let btn = document.createElement("button");
+            btn.innerHTML = `count is ${data.value}`;
+            btn.addEventListener("click", () => {
+                data.value += 1
+            });
+            data.watch(() => {
+                btn.innerHTML = `count is ${data.value}`;
+            }, 'value')
+            return btn
+        }
+        """
+
+    CounterButton().servable()
+    '''
+
+
+class ReactComponent(ReactiveESM):
+    '''
+    The `ReactComponent` allows you to create custom Panel components
+    using React without the complexities of Javascript build tools.
+
+    A `ReactComponent` subclass provides bi-directional syncing of its
+    parameters with arbitrary HTML elements, attributes and
+    properties. The key part of the subclass is the `_esm`
+    variable. Use this to define a `render` function as shown in the
+    example below.
+
+    import panel as pn
+    import param
+
+    pn.extension()
+
+    class CounterButton(ReactComponent):
+
+        value = param.Integer()
+
+        _esm = """
+        export function render({ data, state }) {
+          return (
+            <>
+              <button onClick={() => { data.value += 1 }}>{state.value}</button>
+            </>
+          )
+        }
+        """
+
+    CounterButton().servable()
+    '''
+
+
+    _bokeh_model = _BkReactComponent
+
+    _react_version = '18.2.0'
+
+    def _init_params(self) -> dict[str, Any]:
+        params = super()._init_params()
+        params['react_version'] = self._react_version
+        return params
+
+    def _process_importmap(self):
+        imports = self._importmap.get('imports', {})
+        imports_with_deps = {}
+        for k, v in imports.items():
+            if '?' not in v and not v.endswith('/'):
+                v += f'?deps=react@{self._react_version},react-dom@{self._react_version}'
+            imports_with_deps[k] = v
+        return {
+            'imports': imports_with_deps,
+            'scopes': self._importmap.get('scopes', {})
+        }
+
+class AnyWidgetComponent(ReactComponent):
```

Comment:
I would recommend renaming to `AnyWidget`. I've started wrting `AnyWidgetComponent` in the documentation and it feels repetitive and cumbersome.

Reply:
I've started getting used to the `AnyWidgetComponent` name. I still thinks its cumbersome. But maybe there is also a point i differentiating it from the original `AnyWidget` class and keeping it in line with `JSComponent` and `ReactComponent`?

Reply:
Yeah that's also my feeling. I am similarly ambivalent, Widget and Component feel a little redundant but I do want consistency and `AnyComponent` is probably not sufficiently reminiscent of AnyWidget to really make the association.